plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -1575,19 +1581,24 @@ public TrinoS3StreamingOutputStream(
 
             this.s3 = requireNonNull(s3, "s3 is null");
 
-            this.buffer = new byte[partSize];
+            this.targetPartSize = max(partSize, MINIMAL_TARGET_PART_SIZE);
+            this.initialPartSize = min(INITIAL_BUFFER_SIZE, partSize);
 
             this.bucketName = requireNonNull(bucketName, "bucketName is null");
             this.key = requireNonNull(key, "key is null");
             this.requestCustomizer = requireNonNull(requestCustomizer, "requestCustomizer is null");
             this.uploadIdFactory = requireNonNull(uploadIdFactory, "uploadIdFactory is null");
             this.uploadExecutor = requireNonNull(uploadExecutor, "uploadExecutor is null");
+            this.buffer = new byte[0];
         }
 
         @Override
         public void write(int b)
                 throws IOException
         {
+            if (!initialized) {
```

Comment:
use
```
ensureExtraBytesCapacity(1);
```
where:
```
private void ensureExtraBytesCapacity(int extraBytesCapacity) {
    int totalBytesCapacity = bufferSize + extraBytesCapacity;
    if (buffer.length < totalBytesCapacity) {
       // buffer length might be 0
       int newBytesLength = max(buffer.length, INITIAL_BUFFER_SIZE);
       if (totalBytesCapacity > newBytesLength) {
          // grow array by 50%
          newBytesLength = min(newBytesLength + (newBytesLength >> 1), partSize);
          newBytesLength = max(totalBytesCapacity, newBytesLength);
       }
       buffer = Arrays.copyOf(buffer, newBytesLength);
   }
}
```
see `io.trino.operator.output.SlicePositionsAppender#ensureExtraBytesCapacity` for example